Added adding item feature + remove item feature

ClaraChaouat opened the pull request:
Hi @sandrina-p,

## Summary 
I finally got there with the repo and the app building.
I started to build the grocery app using svelte,
I created a collection in the cloud firestore,
I added an "add items" feature to the app,
I created a "remove items" feature.

## Learnings
- Creating a database with cloud firestore,
- Linking (binding) the database to the svelte app,
- Adding items to the database,
- Removing items from the database.
- How to enable prettier for svelte.

## Struggles
- I struggled creating the app and the repo but I think I'm good now. I wrote all the steps I need to create a repo and an app so I think I'll be fine.
- I still need to investigate on the methods we use to deal with firebase like (db.collection("grocery").doc(itemId).delete()) to understand how it really works. For example, why ".doc" and why ".collection"?
- I still have to investigate on how to edit a list item.
- What is convenient is that I think we can manage sorting directly on the database.
 
## References
- https://www.youtube.com/watch?v=L39XLne7EU8&t=319s 
- firebase documentation

Thanks for your feeback.

ClaraChaouat committed:
Added adding item feature + remove item feature

sandrina-p commented on src/main.js:
```diff
@@ -0,0 +1,10 @@
+import App from './App.svelte';
```
Hi @ClaraChaouat,

It looks like some files are missing from git, did you commit and push everything?

Some files missing: `package.json`.

In these type of projects that require some setup, it's a common practice to leave instructions in the `README.md`. You can look for examples out there of nice readmes. For yours a description and instructions are enough for starting :)

Here are some [awesome readme's](https://github.com/matiassingers/awesome-readme) from the community

ClaraChaouat replied:
Hi Sandrina, 

I was wondering what files you needed to have a look at the project. So, I got that you need the package.json file. Do you need another one? All of them maybe? I didn't want to push all of them because I wasn't sure I was relevant.

sandrina-p replied:
In a project usually, you commit all the files, except `node_modules` and cache folders `ex __svelte__` :)

ClaraChaouat committed:
Added package.json file

ClaraChaouat committed:
Added all the missing files

ClaraChaouat committed:
Added missing files

ClaraChaouat committed:
Added prettier formatting

ClaraChaouat commented:
Hi @sandrina-p , 

Normally, you should find all the files needed.

Thanks for your feedback :) .

ClaraChaouat committed:
Added favicon file

ClaraChaouat committed:
Update README.md

ClaraChaouat committed:
Update README.md

ClaraChaouat commented:
Hi @sandrina-p ,

I added in the README.md file the instructions to start the project.
Let me know If it's ok.

Thank you.

ClaraChaouat committed:
Updated README.file

sandrina-p commented on src/App.svelte:
```diff
@@ -0,0 +1,72 @@
+
+<main>
+  <h1>Grocery list</h1>
+  <h2>
```
This is invalid HTML, H2 does not support form controls (eg input) inside.

ClaraChaouat replied:
What does "H2 does not support form controls (eg input) inside." mean?

sandrina-p replied:
It means you cannot have elements like input inside h2, as you did. It needs to be separated elements. ex:

```
<h2>this is a title</h2>
<input type="text" />
```

ClaraChaouat replied:
Oh Ok that's good to know.

ClaraChaouat committed:
Merge branch 'working-branch' of https://github.com/ClaraChaouat/svelte into working-branch

sandrina-p commented on src/App.svelte:
```diff
@@ -0,0 +1,72 @@
+      type="number"
+      bind:value={newItemQuantity}
+      placeholder="Quantity to buy"
+    />
```
This is a form, so it should be wrapped in a `<form>` element and should have labels connected.

Reference: https://www.w3.org/WAI/tutorials/forms/labels/

sandrina-p commented on src/App.svelte:
```diff
@@ -0,0 +1,72 @@
+    {#each arrList as listItem}
+      <li>
+        {listItem.data().name} : {listItem.data().quantity}
+        <button on:click={() => removeItem(listItem.id)}>Remove Item</button>
```
Well done!

sandrina-p commented on src/App.svelte:
```diff
@@ -0,0 +1,72 @@
+
+  @media (min-width: 640px) {
+    main {
+      max-width: none;
```
Are you planing to add more styles or just keep it simple like this?

ClaraChaouat committed:
Correcting README.md file

sandrina-p commented on src/App.svelte:
```diff
@@ -0,0 +1,72 @@
+  let newItemText = "";
+  let newItemQuantity;
+
+  db.collection("grocery").onSnapshot((snapData) => {
```
Each time this component re-renders, a new db collection event is created. This should happen only once when the component mounts. You can use svete `onMount` lifecycle method.

ClaraChaouat replied:
Could you please give me more details what that? I was also confused on that topic with React.
What does "re-render" exactly mean? When does a component "re-render"?
When does the component mount?
For example, what does happen when I add an item to the list? Does the component mount or re-render?

sandrina-p reviewed (commented):
Yeah, congrats on your first svelte app with firebase! 🎉

ClaraChaouat committed:
Added the form tag

ClaraChaouat commented:
hi @sandrina-p ,

A quick question... Adding the <form> tag created something weird in the app. When you add an item to the list (after adding the first element, it removes the item you just added). 
I was not able to find any info about that.
Thanks for your feedback.

ClaraChaouat committed:
Added update items feature

ClaraChaouat commented:
Hi @sandrina-p , 

Here is the last version of the grocery app:

## Summary 
- I made the changes you adviced me to do,
- I added the onMount() method,
- I added an "Edit" feature. When you hover an item, a message displays indicating that you can edit the entry by clicking on it,
- I "splitted" the app into components,
- I added an hover effect on list items to inform users how to edit the item.
## Learnings
-  I learnt about the Edit mode,
- If/ else conditions in Svelte,
- Update() method with Svelte,
- Tooltip feature

## Struggles
- I'd like to do something I couldn't figure out how to do that. So far, when I want to edit an item, I click on it and it disappears. When updating, I'd like the item to stay in the input field. Why that? For example, If the user only wants to update the quantity of the item to buy, He/she wouldn't have to rewrite everything, only the quantity. Does that make sense to you?
- I left a question about your latest comment regarding the onMount () :) .
## References 
- Svelte documentation.

Thank you for your feedback :) .

ClaraChaouat committed:
Added features + design + edit features

ClaraChaouat commented:
Hi @sandrina-p ,

## Summary 
- I started to work on the design,
- I added a date feature (I thought It would be relevant for that kind of app),
- I added some buttons at the bottom of the page which might be useful for that kind of app. I'd like to implement the corresponding features If it's not too complicated (I'd like not to spend too much time on that project),
- I started to change the "Update" feature but I'm struggling trying to do something I'll explain in the "Struggle" session of this PR.
## Learnings
- How to use the date() method with svelte,
- Adding images with svelte.
## Struggles
- I'm struggling with the "update" feature. I'm not able to figure out how to change the input text by clicking the update button.

- Question from a previous PR: could you please give me more details about the mount() method? I was also confused on that topic with React.
What does "re-render" exactly mean? When does a component "re-render"?
When does the component mount?
For example, what does happen when I add an item to the list? Does the component mount or re-render?

- What do you think about implementing the features corresponding to the buttons I added at the bottom of the page? Do you think It's easily doable?
## References 
- Svelte documentation.

ClaraChaouat committed:
Added remove & update item logic - Styling - Responsive design

ClaraChaouat commented:
Hi @sandrina-p ,

## Summary 
- I added the logic for the remove and update buttons. I didn't really understand what you showed me in the video so I tried to find another way to implement the logic.
- I also added some design and animations, especially on the buttons.
- I added the responsive design.
## Learnings
- How to pass a function as a props with svelte,
- How to "build" responsive design.
- Wow moment: when I finally found a way to implement the logic the remove and update function!!!!! I sweat (a lot) :).
 ## Struggles
- Implementing the logic for the remove and update button.
## References 
https://web.dev/responsive-web-design-basics/ This ressource is amazing for the responsive web design fundamentals. Thanks!

Thanks for your feedback :) .

ClaraChaouat committed:
Update README.md

ClaraChaouat committed:
Update README.md

ClaraChaouat committed:
Update README.md

sandrina-p commented on src/App.svelte:
```diff
@@ -0,0 +1,216 @@
+  }
+  ul {
+    list-style-type: none;
+    width: 350px;
```
The page breaks on mobile (320px), it's missing here a `max-width` to make sure it's fully responsive. [see screenshot](https://user-images.githubusercontent.com/14869087/112058382-72879780-8b52-11eb-9360-cccd540d1cb5.png)

ClaraChaouat replied:
@sandrina-p Thanks for this feedback. I was wondering why you talked about the mobile size of 320px and I realized that It is now the smallest mobile size:) .

sandrina-p commented on README.md:
```diff
@@ -0,0 +1,62 @@
+##  :memo: Description
```
It's missing the `.gitignore` file, you need to commit and push it as well, otherwise some files like `node_modules` would be listed as created/changed on my computer because I don't have that file.

ClaraChaouat replied:
@sandrina-p No worries, I didn't know that. I'll do it.

sandrina-p commented on README.md:
```diff
@@ -0,0 +1,62 @@
+## :see_no_evil::hear_no_evil::speak_no_evil: Feedback
+I'm still in a learning process of web development so feel free to shoot me an e-mail If you have any feedbacks/comments or any usefull things that will help me to keep improving :relaxed:.
+
+:email: clara.chaouat@gmail.com
```
This readme is cool! 🚀 well done!

ClaraChaouat replied:
@sandrina-p Thanks. I'm glad you liked it.

sandrina-p commented on src/App.svelte:
```diff
@@ -0,0 +1,216 @@
+    {#each groceryList as listItem}
+      <GroceryItem
+        item={listItem.data()}
+        {removeItem}
```
I must confess I don't know this sugar syntax to add events! Where did you see it to use like this instead of `removeItem={removeItem}`?

sandrina-p replied:
Did you struggled to use follow the tutorial to [passing props as events](https://svelte.dev/tutorial/component-events)?

ClaraChaouat replied:
@sandrina-p No I didn't really understand the logic.

sandrina-p commented on src/components/GroceryItem.svelte:
```diff
@@ -0,0 +1,155 @@
+
+<li class="item">
+  {#if isEditing}
+    <input class="input-update" type="text" bind:value={inputEl} />
```
with `export let inputEl` you are overcomplicating. you don't need to export that as a prop, it can be a simple variable without the export.

What you are looking for is the `value` HTML property to set a initial default value. 

```
<input value={item.name} class="input-update" type="text" bind:value={inputEl} />
```

sandrina-p commented on src/components/GroceryItem.svelte:
```diff
@@ -0,0 +1,155 @@
+    <button
+      class="edit-item-btn btn-white btn-animate"
+      on:click={() => {
+        editItem();
```
When one item is set as `isEditing`, all the items enter in that mode. Didn't you prefer the "isEditing" to be individual per item?

ClaraChaouat replied:
@sandrina-p Yes I think It'd be more relevant.

sandrina-p commented on src/components/GroceryItem.svelte:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  :focus {
+    outline-width: 0px;
```
I have a challenge for you: Try to add a item to the grocery using your keyboard only. No mouse. Use the keys `Tab` and `Shift + Tab` to navigate.

You won't be able to do that because you are removing the focus styles. With this you are limiting those people who use a keyboard, from using your website. Never, ever do that. Please fix it and remember keyboard navigation, always, in your future projects.

Reference about focus styles:
- https://css-tricks.com/focusing-on-focus-styles/
- https://darekkay.com/blog/accessible-focus-indicator/

ClaraChaouat replied:
@sandrina-p Thanks for that info. I wouldn't ever think about that by myself!

ClaraChaouat replied:
Great ressources. Thanks.

sandrina-p commented on src/App.svelte:
```diff
@@ -0,0 +1,216 @@
+    console.log(itemId);
+    console.log(inputEl);
+    console.log(newItemList);
+    db.collection("grocery").doc(itemId).update({ name: inputEl });
```
Maybe in this project is already too much but I highly recommend you to learn about asynchronous code and how to handle it properly.

This youtuber FunfunFunction has great videos about JavaScript stuff. Please, do learn and practice dealing with asyncronous code following his video tutorial:
1. What are Promises: https://www.youtube.com/watch?v=2d7s3spWAzo&list=PL0zVEGEvSaeEd9hlmCXrk5yUyqUag-n84&ab_channel=FunFunFunction
2. Async await: https://www.youtube.com/watch?v=568g8hxJJp4&t=922s&ab_channel=FunFunFunction

sandrina-p reviewed (commented):
Well done in doing this groceries app 🎉 

I hope you learned a lot from it and can apply the new knowledge in future projects.
I hope one day, you can come back to this project and be able to improve it on your own with your future knowledge :)

sandrina-p commented on src/App.svelte:
```diff
@@ -0,0 +1,216 @@
+<header />
+<main>
+  <h1>Grocery list</h1>
+  <h3>
```
This might look like a heading, but it's not, semantically speaking, a heading. It can be a span with CSS.

Reference: https://accessibility.umn.edu/what-you-can-do/start-7-core-skills/headings

ClaraChaouat replied:
@sandrina-p Thanks for the video, that was very interesting. 
To be sure that I got your point, in this case It's more relevant to put a <span> tag instead of an <header> tag and even if I'd still add this header tag I should do it differently. My question is how? Did you mean that It doesn't make sense to put a header tag with nothing in it?

ClaraChaouat replied:
Or maybe you meant that header stylings have to be writing first in the <style> section?

sandrina-p replied:
What I mean is that it doesn't make sense the date to be a h3 because it doesn't hold that purpose. it's just a simple date, not a type of heading. imagine this website was a book, a recipe book, the date wouldn't be a title, right?

so what I meant is to just replace it with a span.

here's another resource, a short video, about headings: https://www.youtube.com/watch?v=vAAzdi1xuUY

ClaraChaouat replied:
Got it!

ClaraChaouat committed:
Commited the gitignore file

ClaraChaouat committed:
Merge branch 'working-branch' of https://github.com/ClaraChaouat/svelte into working-branch

ClaraChaouat committed:
Update README.md

ClaraChaouat committed:
Update README.md

ClaraChaouat committed:
Last changes

ClaraChaouat committed:
Last changes

ClaraChaouat commented:
Hi @sandrina-p ,



## Summary 
- I pushed some changes with the advice you gave me.


## Learnings
- What's the smallest mobile screensize,
- How important styling on focus can be for accessibility.
- What is and how a screenreader works (shame on me I didn't know).

## Struggles
- There are some questions about your last comments.

## References 
The one you gave me.

Thanks for your feedback :) .

ClaraChaouat committed:
Added footer details

ClaraChaouat committed:
Added gif file

ClaraChaouat committed:
Update README.md

ClaraChaouat committed:
Update README.md

sandrina-p commented on src/AddItemForm.svelte:
```diff
@@ -0,0 +1,76 @@
+
+  .item-input:focus {
+    outline: var(--action-btn-color) solid 2px;
+    box-shadow: 0 0 0 2px hsla(210, 100%, 50%, 0.5);
```
@ClaraChaouat the only thing missing is the focus state in each interactive element (the buttons) :)

You can test it by using the keyboard `tab` and `shift + tab` to navigate between the page elements.

ClaraChaouat replied:
@sandrina-p , so in the same logic would you recommend to also add focus to every list items? 
What are the elements we have to pay attention to the focus state?

ClaraChaouat replied:
@sandrina-p is "testing" an app or a website as you used a sreenreader  something you recommend to do for every project? (I guess the answer is yes, but I'm asking anyway :))

sandrina-p commented on src/components/GroceryItem.svelte:
```diff
@@ -0,0 +1,146 @@
+  {/if}
+
+  <button class="remove-item-btn" on:click={() => removeItem(itemId)}
+    ><img src={trashIconImage} alt="{trashIconImage} image" /></button
```
People who use screen readers to navigate a webpage won't know the purpose of this button because it doesn't has text, only icons (in this case an image). 

This image is a [functional image](https://www.w3.org/WAI/tutorials/images/functional/). That means it's used as an action, in this case, to delete the item.

So, the `alt` should reflect that. We can do `alt="Delete product".`

That way, people using screen readers, when they reach this button it will announce `"button, image delete product"`. 

Please review the buttons that are only an image and be mindful about their accessible name :)

ClaraChaouat replied:
@sandrina-p Thanks for the resource, that is good to know :).
